Allow POSTing a single event.

aperiodic opened the pull request:
The collector now allows a single event object to be posted, as well as an array of event objects. Tests are included.

I also fixed a bug in `test.js` where no data would be sent to the server by the `test.request()` method. Previously, the returned closure checked for the presence of a data argument by looking at the value of `arguments.length`, but since it was a closure with no arguments, this didn't quite work. I modified the test to check that the enclosed data argument is truthy and has a length greater than zero.

Commit:
Send request body in tests, if provided.

Previously, the test.request function in test.js checked for a data
argument by looking at the value of `arguments.length`, but since it was
inside a closure with no arguments, the test would never return true,
even if a data argument was passed to the enclosing function.

Now, the test is whether the data argument of the enclosing function is
truthy and has a length greater than zero.

Commit:
Allow POSTing of a single event.

When POSTing event data to the `/1.0/event/put` endpoint, the POSTed
JSON data can either be an array of event objects, or a single event
object. Any other kind of JSON is ignored, and a 400 Malformed Request
response is sent to the client.

Test for a 400 response when POSTing either bad JSON or JSON whose root
value is not an array or object, and a 200 response otherwise.

mbostock commented:
Good catch on the test framework.

I'm a fan of parsimonious APIs: if there's only one way to do something, then there's less to learn (and to implement, document, and test). If you want to post a single event, it seems reasonable to wrap the event in square brackets, sending "[event]" rather than "event". That's only two additional characters. Also, it allows the API to indicate that you can batch multiple events together for greater efficiency.

Commit:
Revert collector changes, keep tests.

aperiodic commented:
That's a fair point on the API. I was just frustrated that the events I was sending–which were valid JSON–elicited 400 responses from the collector, and that I had to dig into the source to figure out the underlying cause.

Since this is really just a documentation issue, I've added a paragraph on sending events to the end of the main wiki page, and added a commit that reverts the collector changes while keeping the new tests (one is modified to ensure that plain events are now rejected).

mbostock commented:
Nice, thank you!
